osu.Server.Spectator/Hubs/MultiplayerHub.cs

```diff
@@ -0,0 +1,358 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Caching.Distributed;
+using osu.Game.Online.RealtimeMultiplayer;
+
+namespace osu.Server.Spectator.Hubs
+{
+    public class MultiplayerHub : StatefulUserHub<IMultiplayerClient, MultiplayerClientState>, IMultiplayerServer
+    {
+        // for the time being rooms will be maintained in memory and not distributed.
+        private static readonly Dictionary<long, MultiplayerRoom> active_rooms = new Dictionary<long, MultiplayerRoom>();
```

Comment:
i switched away from ConcurrentDictionary for that reason specifically. Doesn't seem like the class we want here.

Keeping locking simple until performance is actually a problem (doubt it will ever be).